File: remove-palindromic-subsequences/solution1332.go
// https://leetcode.com/problems/remove-palindromic-subsequences/

package main

import "fmt"

// T: O(n), S: O(1)
func removePalindromeSub(s string) int {
	if s == "" {
		return 0
	}
	if isPalindrome(s) {
		return 1
	}
	return 2
}

func isPalindrome(s string) bool {
	left, right := 0, len(s)-1
	for left < right {
		if s[left] != s[right] {
			return false
		}
		left++
		right--
	}
	return true
}

func main() {
	s1 := "ababa"
	s2 := "abb"
	s3 := "baabb"
	fmt.Println(removePalindromeSub(s1)) // 1
	fmt.Println(removePalindromeSub(s2)) // 2
	fmt.Println(removePalindromeSub(s3)) // 2
}

File: remove-palindromic-subsequences/Solution1332.java
// https://leetcode.com/problems/remove-palindromic-subsequences/

class Solution1332 {

    // T: O(n), S: O(1)
    public int removePalindromeSub(String s) {
        if (s.isEmpty()) {
            return 0;
        }
        if (isPalindrome(s)) {
            return 1;
        }
        return 2;
    }

    private boolean isPalindrome(String s) {
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    public static void main(String[] args) {
        var sol = new Solution1332();
        var s1 = "ababa";
        var s2 = "abb";
        var s3 = "baabb";
        System.out.println(sol.removePalindromeSub(s1)); // 1
        System.out.println(sol.removePalindromeSub(s2)); // 2
        System.out.println(sol.removePalindromeSub(s3)); // 2
    }

}
